runtime/sema/type.go

```diff
@@ -5021,6 +5022,286 @@ func (t *DictionaryType) Resolve(typeArguments *TypeParameterTypeOrderedMap) Typ
 	}
 }
 
+// InclusiveRangeType
+
+type InclusiveRangeType struct {
+	MemberType          Type
+	memberResolvers     map[string]MemberResolver
+	memberResolversOnce sync.Once
+}
+
+var _ Type = &InclusiveRangeType{}
+var _ ParameterizedType = &InclusiveRangeType{}
+
+func NewInclusiveRangeType(memoryGauge common.MemoryGauge, elementType Type) *InclusiveRangeType {
+	common.UseMemory(memoryGauge, common.DictionarySemaTypeMemoryUsage)
+	return &InclusiveRangeType{
+		MemberType: elementType,
+	}
+}
+
+func (*InclusiveRangeType) IsType() {}
+
+func (*InclusiveRangeType) Tag() TypeTag {
+	return InclusiveRangeTypeTag
+}
+
+func (r *InclusiveRangeType) String() string {
+	memberString := ""
+	if r.MemberType != nil {
+		memberString = fmt.Sprintf("<%s>", r.MemberType.String())
+	}
+	return fmt.Sprintf(
+		"InclusiveRange%s",
+		memberString,
+	)
+}
+
+func (r *InclusiveRangeType) QualifiedString() string {
+	return fmt.Sprintf(
+		"InclusiveRange<%s>",
+		r.MemberType.QualifiedString(),
+	)
+}
+
+func (r *InclusiveRangeType) ID() TypeID {
+	return TypeID(fmt.Sprintf(
+		"InclusiveRange<%s>",
+		r.MemberType.ID(),
+	))
+}
+
+func (r *InclusiveRangeType) Equal(other Type) bool {
+	otherRange, ok := other.(*InclusiveRangeType)
+	if !ok {
+		return false
+	}
+	if otherRange.MemberType == nil {
+		return r.MemberType == nil
+	}
```

Comment:
When would the `MemberType` be `nil`?

Reply:
I do not have a definitive answer to this but I'll explain why I added it.

The `InclusiveRangeType` must be declared in the `BaseTypeActivation` in the `init` function of `runtime/sema/type.go`. Since we don't know the inner type, we leave it as nil. `CapabilityType` is implemented the same way.

If I don't declare it in `BaseTypeActivation` then code such as below fails with "error: cannot find type in this scope: InclusiveRange".
```cadence
let r: InclusiveRange<Int> = InclusiveRange(s, e, step: step)
```

Please let me know if there is a better way.

Reply:
Good point, I've updated `InclusiveRangeType` to handle the case of MemberType as nil in 1f96574c0.

Looking into the `checkParameterizedTypeIsInstantiated` suggestion.

Reply:
> We probably have to discuss this a bit more, but I think we can do this in checkTypeAnnotation, which e.g. already checks if an interface type is used as-is (i.e. calls checkInvalidInterfaceAsType). A new function "checkParameterizedTypeIsInstantiated" could check that the elements of TypeArguments() for all ParameterizedType is non-nil for all TypeParameters() that are not Optional.

Seems like we already have this check in place. I added a test case in e6eeaed9d to use InclusiveRange without providing a type argument and received a `InvalidTypeArgumentCountError`.

Reply:
Ah, nice. Yeah, this handles the case of an incorrect, number of type arguments, but only if an instantiation is used (`ast.InstantiationType`). I don't think it handles the case where the generic type is used on its own yet (bare `ast.NominalType`, not wrapped in an `ast.InstantiationType`).

Reply:
Makes sense. Added `checkParameterizedTypeIsInstantiated` in d97eef694.